### PR TITLE
´Use g++ and clang++ instead of gcc and clang as C++ compilers

### PR DESCRIPTION
**Pull request by RCoeurjoly**
This would solve the CXX and LD points of https://github.com/YosysHQ/yosys/issues/2011.
LD is also changed to g++ when appropriate.

All pipelines pass in my fork (https://github.com/RCoeurjoly/yosys/actions).

### CODE REVIEW

**Review by whitequark — CHANGES_REQUESTED** [review]
`clang++` is not a linker, it is a compiler driver, so it should not be used in the `$(LD)` variable. Either put a linker (`ld`, `lld`, etc) in `$(LD)` or remove `$(LD)` altogether and change `$(LDFLAGS)` to `$(LINKFLAGS)` or something.

**Comment by whitequark** [comment]
> I think it is not viable to use ld

That is possible. In that case you should remove `$(LD)` entirely and use `$(CXX)` instead to link.

**Comment by RCoeurjoly** [comment]
> > I think it is not viable to use ld
> 
> That is possible. In that case you should remove `$(LD)` entirely and use `$(CXX)` instead to link.

Done. I had to change $LD for $CXX in passes/techmap/Makefile.inc, otherwise we would get the same linking error

> ld: unrecognised option: -rdynamic

**Comment by whitequark** [comment]
Looks better, thank you. Now that we don't have `$(LD)`, we should not use `$(LDFLAGS)` either; I propose `$(LINKFLAGS)`. Similarly, instead of `$(LDLIBS)` we can use `$(LIBS)`.

It looks like we have a bunch of `$(filter-out -rdynamic)` clauses that can probably be simplified, but I'll take another look at that later, and it doesn't strictly speaking have to be a part of this PR.

**Comment by RCoeurjoly** [comment]
> Looks better, thank you. Now that we don't have `$(LD)`, we should not use `$(LDFLAGS)` either; I propose `$(LINKFLAGS)`. Similarly, instead of `$(LDLIBS)` we can use `$(LIBS)`.
> 
> It looks like we have a bunch of `$(filter-out -rdynamic)` clauses that can probably be simplified by not adding `-rdynamic` by default in first place, but I'll take another look at that later, and it doesn't strictly speaking have to be a part of this PR.

Done. I had to fix it in other Makefiles too.

Now two tests fail

- one in macOS (https://github.com/RCoeurjoly/yosys/actions/runs/8041007987/job/21959774675#step:10:1562) and 
- one in Linux (https://github.com/RCoeurjoly/yosys/actions/runs/8041007983/job/21959774813#step:12:1444).

I have reproduced the error locally by running:
`cd tests/various`
`bash run-test.sh`

I get the following error:

> make: *** [run-test.mk:456: async.sh] Error 127

I am a bit lost as to how to debug that failure.

**Comment by povik** [comment]
> Now two tests fail
> 
>    * one in macOS (https://github.com/RCoeurjoly/yosys/actions/runs/8041007987/job/21959774675#step:10:1562) and
> 
>    * one in Linux (https://github.com/RCoeurjoly/yosys/actions/runs/8041007983/job/21959774813#step:12:1444).

Those happen to be the only CI jobs which run the testsuite (through `make test`), the other CI jobs build the binaries but don't run `make test`.

Looks like the fault is in `tests/various/plugin.sh` because `yosys-config` needs updating, see `misc/yosys-config.in`.

File: misc/yosys-config.in [hunk 1]
```diff
@@ -9,8 +9,8 @@ help() {
 		echo "Replacement args:"
 		echo "    --cxx         @CXX@"
 		echo "    --cxxflags    $( echo '@CXXFLAGS@' | fmt -w60 | sed ':a;N;$!ba;s/\n/ \\\n                      /g' )"
-		echo "    --ldflags     @LDFLAGS@"
-		echo "    --ldlibs      @LDLIBS@"
+		echo "    --ldflags     @LINKFLAGS@"
+		echo "    --ldlibs      @LIBS@"
```
Comment (whitequark) [inline misc/yosys-config.in:13]: I think we should rename them here as well, while providing compatibility:

```suggestion
		echo "    --linkflags   @LINKFLAGS@"
		echo "    --ldflags     (alias of --linkflags)
		echo "    --linklibs    @LIBS@"
		echo "    --ldlibs      (alias of --linklibs)
```

File: libs/ezsat/Makefile [hunk 2]
```diff
@@ -1,9 +1,9 @@
 
 CC = clang
-CXX = clang
+CXX = clang++
 CXXFLAGS = -MD -Wall -Wextra -ggdb
 CXXFLAGS += -std=c++11 -O0
-LDLIBS = ../minisat/Options.cc ../minisat/SimpSolver.cc ../minisat/Solver.cc ../minisat/System.cc -lm -lstdc++
+LIBS = ../minisat/Options.cc ../minisat/SimpSolver.cc ../minisat/Solver.cc ../minisat/System.cc -lm -lstdc++
```
Comment (whitequark) [inline libs/ezsat/Makefile:6]: This is a completely unhinged definition of `LIBS`.

**Review by whitequark — APPROVED** [review]
Looks good to me, thank you a lot! The Makefiles are not *perfect* but they are at least not blatantly misleading anymore, so we are getting somewhere!!

File: libs/ezsat/Makefile (hunk 2 above)
Comment (RCoeurjoly) [reply libs/ezsat/Makefile:6]: I was surprised too.

**Comment by RCoeurjoly** [comment]
> Looks good to me, thank you a lot! The Makefiles are not _perfect_ but they are at least not blatantly misleading anymore, so we are getting somewhere!!

Thank you for your patience and handholding!

**Pull request merged**